libraries/AC_Avoidance/AC_Avoid.cpp

```diff
@@ -882,11 +998,127 @@ void AC_Avoid::adjust_velocity_proximity(float kP, float accel_cmss, Vector2f &d
     if (_proximity.get_status() != AP_Proximity::Status::Good) {
         return;
     }
+    
+    const AP_AHRS &_ahrs = AP::ahrs();
+    
+    // do not adjust velocity if vehicle is outside the polygon fence
+   
+    // Safe_vel will be adjusted to remain within fence.
+    // We need a separate vector in case adjustment fails,
+    // e.g. if we are exactly on the boundary.
+    Vector3f safe_vel(desired_vel_cms);
+    Vector2f desired_back_vel_cms_xy;
 
-    // get boundary from proximity sensor
+    // for backing away
+    Vector2f quad_1_back_vel, quad_2_back_vel, quad_3_back_vel, quad_4_back_vel;
+    float max_back_vel_z = 0.0f;
+    float min_back_vel_z = 0.0f; 
+
+    // if boundary points are in body-frame, rotate velocity vector from earth frame to body-frame
+    Vector2f safe_vel_2d = _ahrs.earth_to_body2D(Vector2f(desired_vel_cms.x,desired_vel_cms.y));
+    safe_vel = Vector3f{safe_vel_2d.x, safe_vel_2d.y, desired_vel_cms.z};
+        
+    // calc margin in cm
+    const float margin_cm = MAX(_margin * 100.0f, 0.0f);
+    Vector3f stopping_point_plus_margin; 
+    if (!desired_vel_cms.is_zero()) {
+        // for stopping
+        const float speed = safe_vel.length();
+        stopping_point_plus_margin = safe_vel*((2.0f + margin_cm + get_stopping_distance(kP, accel_cmss, speed))/speed);
+    }
+    
     uint16_t num_points = 0;
-    const Vector2f *boundary = _proximity.get_boundary_points(num_points);
-    adjust_velocity_polygon(kP, accel_cmss, desired_vel_cms, backup_vel, boundary, num_points, false, _margin, dt, true);
+    uint32_t stack_bit = 0;
+    // get boundary from proximity sensor
+    Vector3f (*boundary)[5] = _proximity.get_boundary_points(num_points, stack_bit);
+    // exit if there are no points
+    if (boundary == nullptr || num_points == 0) {
+        return;
+    }
+
+    for( uint8_t stack = 0; stack < 5 ; stack++) {
+        uint32_t bit = 1<<stack;
+        if (!(bit & stack_bit)) {
+            // no valid points at this pitch
```

Comment:
Probably better not to use `pitch` as we really use that for aircraft attitude.

Perhaps `azimuth`, 'though that usually related to earth-frame rather than body-frame.

Reply:
i'm not sure we should use a new word actually.  The boundary is locked in earth-frame for pitch so I think sticking with "pitch" is ok.